Enable format string attributes in Objective-C++ sources

ddkilzer opened the pull request:
#### 1c37850168d747622bc2bab5909a5d5b5bdd8b73
<pre>
Enable format string attributes in Objective-C++ sources
<a href="https://rdar.apple.com/159245047">rdar://159245047</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=297946">https://bugs.webkit.org/show_bug.cgi?id=297946</a>

Reviewed by Mike Wyrzykowski, Per Arne Vollan, and Ben Nham.

Clang introduced `__attribute__((__format__(__NSString__, format, va_args)))`
a few years ago, which is a superset of the `printf` format attribute
and supports the `%@` placeholder.

Since the assertion and logging code in WTF all support the `%@`
placeholder, switch them to use the `__NSString__` format attribute on
Apple OSes, then fix all the resulting format warnings.

If no comment below, the change is to fix placeholders and/or parameters.

* Source/WTF/wtf/Assertions.cpp:
- Switch logging methods from WTF_ATTRIBUTE_PRINTF to
  WTF_ATTRIBUTE_NSSTRING.
(WTF::createWithFormatAndArguments):
- Extract `formatSpan` into a variable since it&apos;s used with USE(CF), we
  will want to switch to using that in the future anyway, and because it
  makes clang forget that `format` has an `__NSString__` format
  attribute when passing to functions with a `printf` format attribute.
(WTF::vprintf_stderr_common):
- Ditto of comments for WTF::createWithFormatAndArguments().
- Also disable the -Wformat-nonliteral warnings for the whole function
  like WTF::createWithFormatAndArguments().
* Source/WTF/wtf/Assertions.h:
(WTF_ATTRIBUTE_NSSTRING): Add.
- Declare compiler attribute for __NSString__ format strings with a
  fallback to use WTF_ATTRIBUTE_PRINTF() on non-Apple platforms.
(WTF_ATTRIBUTE_PRINTF):
- Remove hack that disables format string attributes in Objective-C++.
- Switch logging functions to use WTF_ATTRIBUTE_NSSTRING() attribute.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
(WTF::LibraryPathDiagnosticsLogger::logError):
- Add missing format string attribute.

* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm:
(WebCore::keyIdentifierForKeyEvent):
* Source/WebCore/platform/ios/wak/WebCoreThread.mm:
(WebCoreObjCDeallocOnWebThread):
(WebCoreObjCDeallocWithWebThreadLock):
* Source/WebCore/platform/gamepad/mac/MultiGamepadProvider.mm:
(WebCore::MultiGamepadProvider::platformGamepadConnected):
(WebCore::MultiGamepadProvider::platformGamepadDisconnected):
* Source/WebCore/platform/mac/PlatformEventFactoryMac.mm:
(WebCore::keyIdentifierForKeyEvent):
* Source/WebCore/platform/network/mac/WebCoreResourceHandleAsOperationQueueDelegate.mm:
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:canAuthenticateAgainstProtectionSpace:]):
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:didReceiveResponse:]):
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:didSendBodyData:totalBytesWritten:totalBytesExpectedToWrite:]):

* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
- Use -localizedDescription instead of logging whole `error` object.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:willPerformHTTPRedirection:newRequest:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:_schemeUpgraded:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:didReceiveChallenge:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):
(-[WKNetworkSessionDelegate URLSession:task:didFinishCollectingMetrics:]):
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
(WebKit::NetworkSessionCocoa::removeNetworkWebsiteData):
- Switch LOG() (which doesn&apos;t support the {public} modifier) to
  RELEASE_LOG_DEBUG() (which does support it, and is not enabled by
  default).
* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::hasProhibitedUsageStrings):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::ensureSandboxCacheDirectory):
* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::checkSandboxRequirementForType):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/WKSOAuthorizationDelegate.mm:
(-[WKSOAuthorizationDelegate authorization:didCompleteWithError:]):
* Source/WebKit/UIProcess/Cocoa/WebInspectorPreferenceObserver.mm:
(-[WKWebInspectorPreferenceObserver init]):
- Move bridge_cast() to create RetainPtr&lt;NSString&gt;.
- Fix placeholder for `NSString *` argument.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm:
(-[WKDigitalCredentialsPicker handleNSError:]):
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.mm:
(WebKit::MediaUsageManagerCocoa::updateMediaUsage):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::clearAllCredentials):
(WebKit::LocalAuthenticator::receiveException const):
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(-[WKMouseInteraction beginPointerLockMouseTracking]):
- Switch to use PRIVATE_LOG_STRING.
- Switch LOG() (which doesn&apos;t support the {private} modifier) to
  RELEASE_LOG_DEBUG() (which does support it, and is not enabled by
  default).
* Source/WebKit/UIProcess/mac/WKPrintingView.mm:
(-[WKPrintingView _preparePDFDataForPrintingOnSecondaryThread]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::insertText):
(WebKit::WebViewImpl::selectedRangeWithCompletionHandler):
(WebKit::WebViewImpl::markedRangeWithCompletionHandler):
(WebKit::WebViewImpl::attributedSubstringForProposedRange):
(WebKit::WebViewImpl::firstRectForCharacterRange):
(WebKit::WebViewImpl::characterIndexForPoint):
(WebKit::WebViewImpl::setMarkedText):
* Source/WebKit/WebProcess/InjectedBundle/mac/InjectedBundleMac.mm:
(WebKit::InjectedBundle::extendClassesForParameterCoder):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createPreviewsForModelWithURL):
(WebKit::ARKitInlinePreviewModelPlayerMac::didCreateRemotePreviewForModelWithURL):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView firstRectForCharacterRange:]):
(-[WebHTMLView ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebHTMLView attributedSubstringFromRange:]):
(-[WebHTMLView setMarkedText:selectedRange:]):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView _menuItemsFromPDFKitForEvent:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView addObserver:forKeyPath:options:context:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
(generateHtml):
- Add ALLOW_NONLITERAL_FORMAT_BEGIN/ALLOW_NONLITERAL_FORMAT_END macros
  to silence nonliteral format warnings.  All of the values passed into
  this function come from `static const char*` globals.

Canonical link: <a href="https://commits.webkit.org/299238@main">https://commits.webkit.org/299238@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c28c4106ca4686155c24f37c5f7f80bfd0e2d816

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118331 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38011 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124494 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70383 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/00f6498a-d170-4fde-be0e-a866e70878cb) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38706 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46593 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89803 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59421 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2cb9e9d5-0a32-409d-8d88-ac564862a23c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121284 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106104 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70292 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/dba9d126-e120-419e-bfd0-a50ca1a3ac60) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29894 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24213 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68157 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/110446 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100265 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24403 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127566 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116843 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45237 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34112 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98481 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45600 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102325 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98268 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43672 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21658 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41713 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18856 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45107 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50783 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/145541 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44570 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37439 "Found 1 new JSC binary failure: testapi, Found 19704 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/push2.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/shift_unshift.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/blue_1086262.js.default, ChakraCore.yaml/ChakraCore/test/Bugs/bug215238.shrua-2.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47914 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46257 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pvollan commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -317,7 +317,7 @@ static uint32_t computeAppleGPUFamily(id<MTLDevice> device)
         id<MTLLibrary> library = [device newLibraryWithSource:@"[[vertex]] float4 vsNop() { return (float4)0; }" options:options error:&error];
         if (error)
-            WTFLogAlways("%@", error); // NOLINT
+            WTFNSLogAlways("%@", error); // NOLINT
```
I wonder if we should just log the localized description of the error as a string here, instead of logging the object.

mwyrzykowski replied:
That would likely be better, though realistically we could also release assert here as `[[vertex]] float4 vsNop() { return (float4)0; }` should always compile

ddkilzer replied:
Will fix

bnham reviewed (approved):
LGTM pending tests

Commit:
Enable format string attributes in Objective-C++ sources
rdar://159245047
https://bugs.webkit.org/show_bug.cgi?id=297946

Reviewed by Mike Wyrzykowski, Per Arne Vollan, and Ben Nham.

Clang introduced `__attribute__((__format__(__NSString__, format, va_args)))`
a few years ago, which is a superset of the `printf` format attribute
and supports the `%@` placeholder.

Since the assertion and logging code in WTF all support the `%@`
placeholder, switch them to use the `__NSString__` format attribute on
Apple OSes, then fix all the resulting format warnings.

If no comment below, the change is to fix placeholders and/or parameters.

* Source/WTF/wtf/Assertions.cpp:
- Switch logging methods from WTF_ATTRIBUTE_PRINTF to
  WTF_ATTRIBUTE_NSSTRING.
(WTF::createWithFormatAndArguments):
- Extract `formatSpan` into a variable since it's used with USE(CF), we
  will want to switch to using that in the future anyway, and because it
  makes clang forget that `format` has an `__NSString__` format
  attribute when passing to functions with a `printf` format attribute.
(WTF::vprintf_stderr_common):
- Ditto of comments for WTF::createWithFormatAndArguments().
- Also disable the -Wformat-nonliteral warnings for the whole function
  like WTF::createWithFormatAndArguments().
* Source/WTF/wtf/Assertions.h:
(WTF_ATTRIBUTE_NSSTRING): Add.
- Declare compiler attribute for __NSString__ format strings with a
  fallback to use WTF_ATTRIBUTE_PRINTF() on non-Apple platforms.
(WTF_ATTRIBUTE_PRINTF):
- Remove hack that disables format string attributes in Objective-C++.
- Switch logging functions to use WTF_ATTRIBUTE_NSSTRING() attribute.
* Source/WTF/wtf/darwin/LibraryPathDiagnostics.mm:
(WTF::LibraryPathDiagnosticsLogger::logError):
- Add missing format string attribute.

* Source/WebCore/platform/ios/PlatformEventFactoryIOS.mm:
(WebCore::keyIdentifierForKeyEvent):
* Source/WebCore/platform/ios/wak/WebCoreThread.mm:
(WebCoreObjCDeallocOnWebThread):
(WebCoreObjCDeallocWithWebThreadLock):
* Source/WebCore/platform/gamepad/mac/MultiGamepadProvider.mm:
(WebCore::MultiGamepadProvider::platformGamepadConnected):
(WebCore::MultiGamepadProvider::platformGamepadDisconnected):
* Source/WebCore/platform/mac/PlatformEventFactoryMac.mm:
(WebCore::keyIdentifierForKeyEvent):
* Source/WebCore/platform/network/mac/WebCoreResourceHandleAsOperationQueueDelegate.mm:
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:canAuthenticateAgainstProtectionSpace:]):
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:didReceiveResponse:]):
(-[WebCoreResourceHandleAsOperationQueueDelegate connection:didSendBodyData:totalBytesWritten:totalBytesExpectedToWrite:]):

* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
- Use -localizedDescription instead of logging whole `error` object.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(-[WKNetworkSessionDelegate URLSession:task:willPerformHTTPRedirection:newRequest:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:_schemeUpgraded:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:didReceiveChallenge:completionHandler:]):
(-[WKNetworkSessionDelegate URLSession:task:didCompleteWithError:]):
(-[WKNetworkSessionDelegate URLSession:task:didFinishCollectingMetrics:]):
(-[WKNetworkSessionDelegate URLSession:dataTask:didReceiveResponse:completionHandler:]):
(WebKit::NetworkSessionCocoa::removeNetworkWebsiteData):
- Switch LOG() (which doesn't support the {public} modifier) to
  RELEASE_LOG_DEBUG() (which does support it, and is not enabled by
  default).
* Source/WebKit/Shared/Cocoa/DefaultWebBrowserChecks.mm:
(WebKit::hasProhibitedUsageStrings):
* Source/WebKit/Shared/mac/AuxiliaryProcessMac.mm:
(WebKit::ensureSandboxCacheDirectory):
* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::checkSandboxRequirementForType):
* Source/WebKit/UIProcess/Cocoa/SOAuthorization/WKSOAuthorizationDelegate.mm:
(-[WKSOAuthorizationDelegate authorization:didCompleteWithError:]):
* Source/WebKit/UIProcess/Cocoa/WebInspectorPreferenceObserver.mm:
(-[WKWebInspectorPreferenceObserver init]):
- Move bridge_cast() to create RetainPtr<NSString>.
- Fix placeholder for `NSString *` argument.
* Source/WebKit/UIProcess/DigitalCredentials/WKDigitalCredentialsPicker.mm:
(-[WKDigitalCredentialsPicker handleNSError:]):
* Source/WebKit/UIProcess/Media/cocoa/MediaUsageManagerCocoa.mm:
(WebKit::MediaUsageManagerCocoa::updateMediaUsage):
* Source/WebKit/UIProcess/WebAuthentication/Cocoa/LocalAuthenticator.mm:
(WebKit::LocalAuthenticator::clearAllCredentials):
(WebKit::LocalAuthenticator::receiveException const):
* Source/WebKit/UIProcess/ios/WKMouseInteraction.mm:
(-[WKMouseInteraction beginPointerLockMouseTracking]):
- Switch to use PRIVATE_LOG_STRING.
- Switch LOG() (which doesn't support the {private} modifier) to
  RELEASE_LOG_DEBUG() (which does support it, and is not enabled by
  default).
* Source/WebKit/UIProcess/mac/WKPrintingView.mm:
(-[WKPrintingView _preparePDFDataForPrintingOnSecondaryThread]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::insertText):
(WebKit::WebViewImpl::selectedRangeWithCompletionHandler):
(WebKit::WebViewImpl::markedRangeWithCompletionHandler):
(WebKit::WebViewImpl::attributedSubstringForProposedRange):
(WebKit::WebViewImpl::firstRectForCharacterRange):
(WebKit::WebViewImpl::characterIndexForPoint):
(WebKit::WebViewImpl::setMarkedText):
* Source/WebKit/WebProcess/InjectedBundle/mac/InjectedBundleMac.mm:
(WebKit::InjectedBundle::extendClassesForParameterCoder):
* Source/WebKit/WebProcess/Model/mac/ARKitInlinePreviewModelPlayerMac.mm:
(WebKit::ARKitInlinePreviewModelPlayerMac::createPreviewsForModelWithURL):
(WebKit::ARKitInlinePreviewModelPlayerMac::didCreateRemotePreviewForModelWithURL):
* Source/WebKitLegacy/mac/WebView/WebHTMLView.mm:
(-[WebHTMLView firstRectForCharacterRange:]):
(-[WebHTMLView ALLOW_DEPRECATED_IMPLEMENTATIONS_END]):
(-[WebHTMLView attributedSubstringFromRange:]):
(-[WebHTMLView setMarkedText:selectedRange:]):
* Source/WebKitLegacy/mac/WebView/WebPDFView.mm:
(-[WebPDFView _menuItemsFromPDFKitForEvent:]):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView addObserver:forKeyPath:options:context:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm:
(generateHtml):
- Add ALLOW_NONLITERAL_FORMAT_BEGIN/ALLOW_NONLITERAL_FORMAT_END macros
  to silence nonliteral format warnings.  All of the values passed into
  this function come from `static const char*` globals.

Canonical link: https://commits.webkit.org/299238@main
